### ISSUE TITLE
error on ContainerCluster status: Update call failed: the desired mutation for the following field(s) is invalid: [nodeConfig.0.MachineType]

### ISSUE THREAD

**Issue by jlgeering** [GoogleCloudPlatform/k8s-config-connector#277]
**Describe the bug**
The status for ContainerCluster is complaining about an invalid nodeConfig value, although nodeConfig (which I think is optional) was never specified.

**ConfigConnector Version**
1.20.1

**To Reproduce**
*Steps to reproduce the behavior:*

not sure, but I this is more or less what we did:

* created a ContainerCluster with an older KCC version (without a nodeConfig)
* set the deletetion policy to abandon
* probably deleted and recreated the ContainerCluster resource (without deleting the cluster) - not sure if it is relevant
* upgraded to KCC to 1.20.0 ( I believe the issue appeared then )
* handle CVE-2020-14386
  * upgrade master to 1.16.13-gke.401
  * created a new ContainerNodePool, to work around #260
  * deleted the other (non-managed) nodepool, using the console
* upgraded KCC to 1.20.1 - no change

manifest:

```yaml
apiVersion: container.cnrm.cloud.google.com/v1beta1
kind: ContainerCluster
metadata:
  annotations:
    cnrm.cloud.google.com/deletion-policy: abandon
    cnrm.cloud.google.com/remove-default-node-pool: 'true'
  name: <REDACTED>
  namespace: <REDACTED>
spec:
  authenticatorGroupsConfig:
    securityGroup: gke-security-groups@<REDACTED>
  initialNodeCount: 1
  ipAllocationPolicy:
    clusterIpv4CidrBlock: ''
  location: <REDACTED>
  loggingService: logging.googleapis.com/kubernetes
  maintenancePolicy:
    recurringWindow:
      endTime: '2020-01-01T04:30:00Z'
      recurrence: 'FREQ=WEEKLY;BYDAY=MO,TU,WE,TH'
      startTime: '2020-01-01T00:30:00Z'
  masterAuth:
    clientCertificateConfig:
      issueClientCertificate: false
  monitoringService: monitoring.googleapis.com/kubernetes
  releaseChannel:
    channel: REGULAR
  workloadIdentityConfig:
    identityNamespace: <REDACTED>.svc.id.goog
```

status from live manifest:

```yaml
status:
  conditions:
    - lastTransitionTime: '2020-09-15T14:27:08Z'
      message: >-
        Update call failed: the desired mutation for the following field(s) is
        invalid: [nodeConfig.0.MachineType]
      reason: UpdateFailed
      status: 'False'
      type: Ready
  endpoint: <REDACTED>
  instanceGroupUrls: 
    - <REDACTED>
    - <REDACTED>
  labelFingerprint: 14cb7bfd
  masterVersion: 1.16.13-gke.401
  servicesIpv4Cidr: <REDACTED>
```

**Comment by spew** [comment]
Hello @jlgeering thanks for reporting this issue, we will look into this and let you know here when it is fixed.

**Comment by kibbles-n-bytes** [comment]
Hey @jlgeering , we determined that this behavior happens because in the one-node pool case, GKE will have the cluster's node config dynamically mirror that node pool's node config on GET, despite it being un-updatable by the customer. We have added functionality in KCC 1.24.0 to be able to handle this case (and similar ones) more effectively. Could you update to this version, and then abandon and reacquire the `ContainerCluster` object by using our `cnrm.cloud.google.com/deletion-policy: abandon` annotation? And then let us know if you continue to see issues here.

**Comment by jlgeering** [comment]
yup, cannot reproduce anymore, seems fixed

thx!

**Comment by jlgeering** [comment]
hi @kibbles-n-bytes not 100% sure but maybe this is happening again: we switched from a cluster with only 1 node pool with preemptible: true to another node pool without preemptible

```
apiVersion: container.cnrm.cloud.google.com/v1beta1
kind: ContainerCluster
metadata:
  annotations:
    cnrm.cloud.google.com/deletion-policy: abandon
    cnrm.cloud.google.com/remove-default-node-pool: "true"
spec:
  initialNodeCount: 1
status:
  conditions:
    - lastTransitionTime: '2020-11-26T10:10:16Z'
      message: >-
        Update call failed: the desired mutation for the following field(s) is
        invalid: [nodeConfig.0.Preemptible]
      reason: UpdateFailed
      status: 'False'
      type: Ready
```

is the fix also supposed to handle that case?

**Comment by maqiuyujoyce** [comment]
Hi @jlgeering , thank you for the reporting the issue. Just for clarification, is the `preemptible` field managed by the KCC resource? And to better understand your problem, could you provide the steps so that we can reproduce the issue?

**Comment by jlgeering** [comment]
steps are more or less the following ones (from memory):

* create a cluster _cluster-1_ (with `cnrm.cloud.google.com/remove-default-node-pool: 'true'`) and a nodepool _pool-a_ with `preemptible: true`
* create a nodepool _pool-b_ with `preemptible: false`
* migrate everything to _pool-b_
* delete _pool-a_

I don't think there were more steps, but ... memory is not 100%

**Comment by maqiuyujoyce** [comment]
Hi @jlgeering , thank you for the steps to reproduce. Is this issue consistent? I couldn't reproduce it with the steps you provided. For the ContainerCluster, did you happen to configure the `nodeConfig.preemptible` field via Config Connector? Would be great if you can share the output of `kubectl get containercluster [cluster-name] -o yaml` so that we can take a look.

Also, just to confirm, are you using Config Connector version 1.24.0?

**Comment by Synehan** [comment]
Got the same issue here, but with nodePool service account:

- KCC version 1.27.1
- Argocd 1.8.1
- GKE 1.16

Step to reproduce:

- Create a cluster without nodeConfig and with the annotation `cnrm.cloud.google.com/remove-default-node-pool: 'true'`
- Create a NodePool A 
- Create a NodePool B
- Delete the NodePool A

I abandonned the cluster, updated KCC to v1.34 and recreated the cluster, but the nodeConfig field seems still created by KCC

**Comment by jlgeering** [comment]
@maqiuyujoyce I can't really say what version we were using at that time, but we tend to track the latest release pretty closely. And yes, we configured everything via Config Connector.

I don't think I still have access to this cluster or that it still is in a state that would be representative => I don't think I can give you what you were asking for.

**Comment by xiaobaitusi** [comment]
Thanks @jlgeering and @Synehan for the repo steps. We will look into this and post an update when we have more information.

**Comment by LiptonB** [comment]
I believe I have just reproduced this on config connector version 1.44.0 following steps very similar to @Synehan. For me it is the oauthScopes that won't sync:
```
'Update call failed: the desired mutation for the following field(s)
      is invalid: [nodeConfig.0.OauthScopes.# nodeConfig.0.OauthScopes.1277378754
      nodeConfig.0.OauthScopes.172152165]'
```

`cnrm.cloud.google.com/version: 1.44.0`
GKE version: 1.18.15-gke.1501

My Steps:
* Create a ContainerCluster without nodeConfig and with the annotation cnrm.cloud.google.com/remove-default-node-pool: 'true'
* Create a ContainerNodePool A with oauthScopes set to X
* Create a ContainerNodePool B with oauthScopes set to Y
* Delete the ContainerNodePool A (I did this _while_ B was being created in GKE, not sure if that's relevant)
* Wait for both ContainerNodePools to reconcile
* Now:
  * ContainerCluster is in `UpdateFailed`
  * `kubectl get containercluster` shows the cluster nodeConfig as having oauthScopes X
  * `gcloud container clusters describe` shows the cluster as having oauthScopes Y

Unsuccessful Workaround:
* Modifying nodeConfig.oauthScopes to Y directly doesn't work because the field is immutable

Successful Workaround:
* Attaching the `cnrm.cloud.google.com/deletion-policy: "abandon"` annotation, deleting the ContainerCluster, and recreating it appears to have fixed the problem. oauthScopes now appears in the ContainerCluster as Y, even though I didn't set it explicitly. I'm going to test on a second cluster that has this problem, will report back if it does not fix the problem there.

**Comment by maqiuyujoyce** [comment]
Hi @LiptonB , thank you for all the details! Sorry that you ran into this tricky issue. I was able to reproduce the issue following your steps.

> * Delete the ContainerNodePool A (I did this while B was being created in GKE, not sure if that's relevant)

Yes, actually this does matter. I tried to delete ContainerNodePool A after B had been successfully created, and I couldn't see any errors. However, when I was creating ContainerNodePool B and deleting ContainerNodePool A at the same time, I got the same error message. This was because the cluster was still using the stale nodeConfig value (the nodeConfig of node pool A) when the new node pool B was trying to update the cluster's nodeConfig. It should be resolved eventually after another reconciliation.

> * Now:
>   * ContainerCluster is in `UpdateFailed`
>   * `kubectl get containercluster` shows the cluster nodeConfig as having oauthScopes X
>   * `gcloud container clusters describe` shows the cluster as having oauthScopes Y

Did you still see the same error after Config Connector tried to reconcile the failed ContainerCluster again? When I was reproducing the issue, after I saw the error message, I waited for another 10 minutes, and eventually the ContainerCluster became "UpToDate". However, if your cluster was not re-reconciled and was not eventually resolved by itself, then it might be some other issue we'll want to look into.

**Comment by tonybenchsci** [comment]
We were about to open a new bug issue, but found this existing issue so I'll paste it here:

KCC is currently failing to manage container clusters, complaining that nodeConfig.0.MachineType is set wrong.
```
    message: 'Update call failed: the desired mutation for the following field(s)
      is invalid: [nodeConfig.0.MachineType]'
    reason: UpdateFailed
    status: "False"
    type: Ready
```
This is interesting because container clusters don’t have nodes (they have node pools that attach to them).

Inspecting the output from the config connector exporter we find that it is storing values for the “first” node pool in the GCP state (this would have changed when we removed khef1): 

`config-connector export https://container.googleapis.com/v1beta1/projects/<REDACTED>/locations/us-east4/clusters/khef`
```
  nodeConfig:
    diskSizeGb: 100
    diskType: pd-standard
    imageType: COS
    machineType: n1-highmem-2
    metadata:
      disable-legacy-endpoints: "true"
    oauthScopes:
    - https://www.googleapis.com/auth/monitoring
    - https://www.googleapis.com/auth/trace.append
    - https://www.googleapis.com/auth/cloud-platform
    - https://www.googleapis.com/auth/logging.write
    serviceAccountRef:
      external: default
    shieldedInstanceConfig:
      enableIntegrityMonitoring: true
    workloadMetadataConfig:
      nodeMetadata: GKE_METADATA_SERVER
```

This now disagrees with what KCC has for this field, which seems to only update on creation of the container cluster:

`bin/kubectl get containercluster khef -n <REDACTED> -o yaml`
```
  nodeConfig:
    diskSizeGb: 100
    diskType: pd-standard
    imageType: COS
    machineType: n1-standard-4
    metadata:
      disable-legacy-endpoints: "true"
    oauthScopes:
    - https://www.googleapis.com/auth/logging.write
    - https://www.googleapis.com/auth/monitoring
    - https://www.googleapis.com/auth/trace.append
    - https://www.googleapis.com/auth/cloud-platform
    serviceAccountRef:
      external: default
    shieldedInstanceConfig:
      enableIntegrityMonitoring: true
    workloadMetadataConfig:
      nodeMetadata: GKE_METADATA_SERVER
```
Unfortunately, this is immutable in KCC! So we have something that mutates on its own in GCP that we cannot mutate in KCC!
for: "gke/overlays/b6i-root/kcc-cluster/b6i-prod/everything.yaml": admission webhook "deny-immutable-field-updates.cnrm.cloud.google.com" denied the request: cannot make changes on immutable fields: [nodeConfig.machineType]

TL; DR: Deleting the first node pool you make with KCC bricks managing the container cluster with KCC right now.

**Comment by tonybenchsci** [comment]
We're seeing once again when deleting the "first" listed ContainerNodepool in a ContainerCluster, that the `status.instanceGroupUrls` list still contains it, and that `spec.nodeConfig` remains the nodepool configs of the first(removed) ContainerNodepool. When we check the gcloud output directly however, both of these are updated as expected. Seems like KCC just needs to more correctly let these fields float and follow GCP? Although it's interesting because the deletion itself was done using KCC (via kubectl apply --prune) without issues.
